internal/controller/group_controller.go

```diff
@@ -69,9 +72,31 @@ func (r *GroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 
 	groupCR := &usernautdevv1alpha1.Group{}
 
-	if err := r.Client.Get(ctx, req.NamespacedName, groupCR); err != nil {
-		r.log.WithError(err).Error("error fetching the group CR")
-		return ctrl.Result{}, err
+	if err := r.Get(ctx, req.NamespacedName, groupCR); err != nil {
+		r.log.WithError(err).Error("Unable to fetch Group CR")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// adding or deleting finalizer based on CR status
+	if groupCR.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(groupCR, groupFinalizer) {
+			controllerutil.AddFinalizer(groupCR, groupFinalizer)
+			if err := r.Update(ctx, groupCR); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(groupCR, groupFinalizer) {
+			if err := r.deleteBackendsTeam(ctx, groupCR); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			controllerutil.RemoveFinalizer(groupCR, groupFinalizer)
+			if err := r.Update(ctx, groupCR); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
```

Comment:
can we rewrite this as

```go
if groupCR.GetDeletionTimestamp() != nil {
		// perform the deletion and return from here
	}

	// add finalizer if not present
```

also remove the finalizer once the deletion is completed. If we remove the finalizer before, CR will be deleted from the cluster even though the deletion fails